engine/lib/phx/src/engine/engine.rs

```diff
@@ -25,7 +27,13 @@ pub struct Engine {
     pub hmgui: HmGui,
     pub input: Input,
     pub exit_app: bool,
-    pub lua: Lua,
+    pub lua: Rf<Lua>,
+}
+
+// This thread local variable contains a ref counted instance of the current Lua VM.
+// This is used by the panic hook to tell the Lua VM to generate backtrace.
+thread_local! {
+    static CURRENT_LUA_CTX: RefCell<Option<Rf<Lua>>> = RefCell::new(None);
```

Comment:
This introduces double RefCell. I'd suggest to use `Rf<Option<Lua>>` instead.

Reply:
Unfortunately we can't avoid a double RefCell here.

The reason is because we need the `Lua` object to be ref counted as we want to maintain a safe reference to it without needing to use pointers, and the only way to do that is with ref counting.

Now, if we were to make this an `Rf<Option<Lua>>`, then that means we need to do two things here:
1) Clone `self.lua` in the Engine to a new instance of `Rf<Lua>`
2) Somehow convert the `Rf<Lua>` to `Rf<Option<Lua>>` whilst still preserving the ref-count (so when the `Rf<Option<Lua>>` is dropped, then the shared ref count for both this and `self.lua` is updated.

I think doing (2) is impossible, because Rc does not support converting the underlying type from Rc<T> to Rc<U> whilst also maintaining the ref count. Therefore, I think we need to stick with the double RefCell here.

Honestly, the double RefCell is not a big deal. We only ever do mutability checks on the outer RefCell inside `Engine::call_lua` (which is only called once per frame), and the inner RefCell (inside Rf) is only ever used inside the panic handler.

Reply:
Yes, you are right, it's not possible to convert `Rf<Lua>` to `Rf<Option<Lua>>` without cloning.
Other idea is to make `CURRENT_LUA_CTX` an 'origin' of `Rf<Lua>` and clone it to the engine on initialization.
In this case we don't need `Option` and can just use `Rf<Lua>` directly in both places.
Probably there can be some other issues with this approach but otherwise it looks doable.